File: packages/python-dal/src/penguin_dal/document/mongodb.py
"""MongoDB document store backend for penguin-dal."""
from __future__ import annotations

from dataclasses import dataclass
from typing import Any


@dataclass(slots=True)
class FindOptions:
    """Options for find operations."""

    limit: int = 0
    skip: int = 0
    sort: list[tuple[str, int]] | None = None


@dataclass(slots=True)
class MongoConfig:
    """Configuration for MongoDB backend."""

    uri: str
    db_name: str
    server_selection_timeout_ms: int = 5000
    connect_timeout_ms: int = 10000
    max_pool_size: int = 100
    tls: bool = False


class MongoDAL:
    """Synchronous MongoDB document store using pymongo."""

    def __init__(self, config: MongoConfig) -> None:
        """Initialize MongoDB backend with given config."""
        try:
            from pymongo import MongoClient
        except ImportError as e:
            raise ImportError(
                "Install pymongo: pip install penguin-dal[mongodb]"
            ) from e

        self.config = config
        self.client = MongoClient(
            config.uri,
            serverSelectionTimeoutMS=config.server_selection_timeout_ms,
            connectTimeoutMS=config.connect_timeout_ms,
            maxPoolSize=config.max_pool_size,
            tls=config.tls,
        )
        self.db = self.client[config.db_name]

    def _objectid_to_str(self, doc: dict[str, Any]) -> dict[str, Any]:
        """Convert ObjectId to string in document."""
        if "_id" in doc:
            from bson import ObjectId

            if isinstance(doc["_id"], ObjectId):
                doc = dict(doc)
                doc["_id"] = str(doc["_id"])
        return doc

    def insert_one(self, collection: str, document: dict[str, Any]) -> str:
        """Insert document into collection, return str(_id)."""
        coll = self.db[collection]
        result = coll.insert_one(document)
        return str(result.inserted_id)

    def find_one(
        self, collection: str, filter: dict[str, Any]
    ) -> dict[str, Any] | None:
        """Find single document matching filter."""
        coll = self.db[collection]
        doc = coll.find_one(filter)
        if doc is None:
            return None
        return self._objectid_to_str(doc)

    def find(
        self,
        collection: str,
        filter: dict[str, Any],
        opts: FindOptions | None = None,
    ) -> list[dict[str, Any]]:
        """Find documents matching filter with options."""
        coll = self.db[collection]
        opts = opts or FindOptions()

        query = coll.find(filter)
        if opts.skip > 0:
            query = query.skip(opts.skip)
        if opts.limit > 0:
            query = query.limit(opts.limit)
        if opts.sort:
            query = query.sort(opts.sort)

        return [self._objectid_to_str(doc) for doc in query]

    def update_one(
        self,
        collection: str,
        filter: dict[str, Any],
        update: dict[str, Any],
        upsert: bool = False,
    ) -> int:
        """Update single document, return modified count."""
        coll = self.db[collection]
        result = coll.update_one(filter, update, upsert=upsert)
        return result.modified_count

    def delete_one(self, collection: str, filter: dict[str, Any]) -> int:
        """Delete single document, return deleted count."""
        coll = self.db[collection]
        result = coll.delete_one(filter)
        return result.deleted_count

    def count(
        self, collection: str, filter: dict[str, Any] | None = None
    ) -> int:
        """Count documents matching filter."""
        coll = self.db[collection]
        filter = filter or {}
        return coll.count_documents(filter)

    def create_index(
        self,
        collection: str,
        keys: list[tuple[str, int]],
        unique: bool = False,
    ) -> None:
        """Create index on collection."""
        coll = self.db[collection]
        coll.create_index(keys, unique=unique)

    def close(self) -> None:
        """Close MongoDB connection."""
        self.client.close()


class AsyncMongoDAL:
    """Async MongoDB document store using motor."""

    def __init__(self, config: MongoConfig) -> None:
        """Initialize async MongoDB backend with given config."""
        try:
            import motor.motor_asyncio
        except ImportError as e:
            raise ImportError(
                "Install motor: pip install penguin-dal[motor]"
            ) from e

        self.config = config
        self.client = motor.motor_asyncio.AsyncMongoClient(
            config.uri,
            serverSelectionTimeoutMS=config.server_selection_timeout_ms,
            connectTimeoutMS=config.connect_timeout_ms,
            maxPoolSize=config.max_pool_size,
            tls=config.tls,
        )
        self.db = self.client[config.db_name]

    def _objectid_to_str(self, doc: dict[str, Any]) -> dict[str, Any]:
        """Convert ObjectId to string in document."""
        if "_id" in doc:
            from bson import ObjectId

            if isinstance(doc["_id"], ObjectId):
                doc = dict(doc)
                doc["_id"] = str(doc["_id"])
        return doc

    async def insert_one(self, collection: str, document: dict[str, Any]) -> str:
        """Insert document into collection, return str(_id)."""
        coll = self.db[collection]
        result = await coll.insert_one(document)
        return str(result.inserted_id)

    async def find_one(
        self, collection: str, filter: dict[str, Any]
    ) -> dict[str, Any] | None:
        """Find single document matching filter."""
        coll = self.db[collection]
        doc = await coll.find_one(filter)
        if doc is None:
            return None
        return self._objectid_to_str(doc)

    async def find(
        self,
        collection: str,
        filter: dict[str, Any],
        opts: FindOptions | None = None,
    ) -> list[dict[str, Any]]:
        """Find documents matching filter with options."""
        coll = self.db[collection]
        opts = opts or FindOptions()

        query = coll.find(filter)
        if opts.skip > 0:
            query = query.skip(opts.skip)
        if opts.limit > 0:
            query = query.limit(opts.limit)
        if opts.sort:
            query = query.sort(opts.sort)

        docs = await query.to_list(None)
        return [self._objectid_to_str(doc) for doc in docs]

    async def update_one(
        self,
        collection: str,
        filter: dict[str, Any],
        update: dict[str, Any],
        upsert: bool = False,
    ) -> int:
        """Update single document, return modified count."""
        coll = self.db[collection]
        result = await coll.update_one(filter, update, upsert=upsert)
        return result.modified_count

    async def delete_one(self, collection: str, filter: dict[str, Any]) -> int:
        """Delete single document, return deleted count."""
        coll = self.db[collection]
        result = await coll.delete_one(filter)
        return result.deleted_count

    async def count(
        self, collection: str, filter: dict[str, Any] | None = None
    ) -> int:
        """Count documents matching filter."""
        coll = self.db[collection]
        filter = filter or {}
        return await coll.count_documents(filter)

    async def create_index(
        self,
        collection: str,
        keys: list[tuple[str, int]],
        unique: bool = False,
    ) -> None:
        """Create index on collection."""
        coll = self.db[collection]
        await coll.create_index(keys, unique=unique)

    async def close(self) -> None:
        """Close MongoDB connection."""
        self.client.close()

File: packages/go-dal/document/mongodb.go
package document

import (
	"context"
	"fmt"
	"time"

	"github.com/penguintechinc/penguin-libs/packages/go-dal"
	"go.mongodb.org/mongo-driver/bson"
	"go.mongodb.org/mongo-driver/bson/primitive"
	"go.mongodb.org/mongo-driver/mongo"
	"go.mongodb.org/mongo-driver/mongo/options"
)

// mongoCollection is the subset of *mongo.Collection operations used by MongoDB.
type mongoCollection interface {
	InsertOne(ctx context.Context, document interface{}, opts ...*options.InsertOneOptions) (*mongo.InsertOneResult, error)
	FindOne(ctx context.Context, filter interface{}, opts ...*options.FindOneOptions) *mongo.SingleResult
	Find(ctx context.Context, filter interface{}, opts ...*options.FindOptions) (*mongo.Cursor, error)
	UpdateOne(ctx context.Context, filter interface{}, update interface{}, opts ...*options.UpdateOptions) (*mongo.UpdateResult, error)
	DeleteOne(ctx context.Context, filter interface{}, opts ...*options.DeleteOptions) (*mongo.DeleteResult, error)
	CountDocuments(ctx context.Context, filter interface{}, opts ...*options.CountOptions) (int64, error)
	Indexes() mongo.IndexView
}

// mongoDatabase is the subset of *mongo.Database operations used by MongoDB.
type mongoDatabase interface {
	Collection(name string, opts ...*options.CollectionOptions) mongoCollection
}

// mongoClientConn is the subset of *mongo.Client used by MongoDB.
type mongoClientConn interface {
	Disconnect(ctx context.Context) error
}

// realMongoDatabase wraps *mongo.Database to implement mongoDatabase.
type realMongoDatabase struct {
	inner *mongo.Database
}

func (r *realMongoDatabase) Collection(name string, opts ...*options.CollectionOptions) mongoCollection {
	return r.inner.Collection(name, opts...)
}

// MongoConfig configures a MongoDB document store.
type MongoConfig struct {
	URI                    string
	DBName                 string
	ServerSelectionTimeout time.Duration
	ConnectTimeout         time.Duration
	MaxPoolSize            uint64
}

// MongoDB implements dal.DocumentStore for MongoDB.
type MongoDB struct {
	conn mongoClientConn
	db   mongoDatabase
	cfg  MongoConfig
}

// NewMongoDB creates a new MongoDB document store.
func NewMongoDB(ctx context.Context, cfg MongoConfig) (*MongoDB, error) {
	if cfg.URI == "" {
		return nil, fmt.Errorf("go-dal: mongodb: %w: connection URI required", dal.ErrInvalidConfiguration)
	}
	if cfg.DBName == "" {
		return nil, fmt.Errorf("go-dal: mongodb: %w: database name required", dal.ErrInvalidConfiguration)
	}

	if cfg.ServerSelectionTimeout == 0 {
		cfg.ServerSelectionTimeout = 5 * time.Second
	}
	if cfg.ConnectTimeout == 0 {
		cfg.ConnectTimeout = 10 * time.Second
	}
	if cfg.MaxPoolSize == 0 {
		cfg.MaxPoolSize = 100
	}

	clientOpts := options.Client().
		ApplyURI(cfg.URI).
		SetServerSelectionTimeout(cfg.ServerSelectionTimeout).
		SetConnectTimeout(cfg.ConnectTimeout).
		SetMaxPoolSize(cfg.MaxPoolSize)

	client, err := mongo.Connect(ctx, clientOpts)
	if err != nil {
		return nil, fmt.Errorf("go-dal: mongodb: connect: %w", err)
	}

	// Verify connection
	if err := client.Ping(ctx, nil); err != nil {
		client.Disconnect(ctx)
		return nil, fmt.Errorf("go-dal: mongodb: ping: %w", err)
	}

	rawDB := client.Database(cfg.DBName)

	return &MongoDB{
		conn: client,
		db:   &realMongoDatabase{inner: rawDB},
		cfg:  cfg,
	}, nil
}

// NewMongoDBWithDatabase creates a MongoDB using injected database/client interfaces (for testing).
func NewMongoDBWithDatabase(conn mongoClientConn, db mongoDatabase, cfg MongoConfig) *MongoDB {
	return &MongoDB{
		conn: conn,
		db:   db,
		cfg:  cfg,
	}
}

// InsertOne inserts a single document into a collection.
func (m *MongoDB) InsertOne(ctx context.Context, collection string, document interface{}) (string, error) {
	coll := m.db.Collection(collection)

	result, err := coll.InsertOne(ctx, document)
	if err != nil {
		return "", fmt.Errorf("go-dal: mongodb: insert: %w", err)
	}

	// Convert ObjectID to hex string
	if oid, ok := result.InsertedID.(primitive.ObjectID); ok {
		return oid.Hex(), nil
	}

	return fmt.Sprintf("%v", result.InsertedID), nil
}

// FindOne finds a single document matching the filter.
func (m *MongoDB) FindOne(ctx context.Context, collection string, filter interface{}) (map[string]interface{}, error) {
	coll := m.db.Collection(collection)

	result := bson.M{}
	err := coll.FindOne(ctx, filter).Decode(&result)
	if err == mongo.ErrNoDocuments {
		return nil, fmt.Errorf("go-dal: mongodb: find: %w", dal.ErrNotFound)
	}
	if err != nil {
		return nil, fmt.Errorf("go-dal: mongodb: find: %w", err)
	}

	// Convert ObjectID to string in result
	if id, ok := result["_id"].(primitive.ObjectID); ok {
		result["_id"] = id.Hex()
	}

	return result, nil
}

// Find finds multiple documents matching the filter.
func (m *MongoDB) Find(ctx context.Context, collection string, filter interface{}, opts ...dal.FindOption) ([]map[string]interface{}, error) {
	coll := m.db.Collection(collection)

	fo := &dal.FindOptions{}
	for _, opt := range opts {
		opt(fo)
	}

	findOpts := options.Find()
	if fo.Limit > 0 {
		findOpts = findOpts.SetLimit(fo.Limit)
	}
	if fo.Skip > 0 {
		findOpts = findOpts.SetSkip(fo.Skip)
	}

	// Apply sort options
	if len(fo.Sort) > 0 {
		sortDoc := bson.D{}
		for _, ik := range fo.Sort {
			sortDoc = append(sortDoc, bson.E{Key: ik.Field, Value: ik.Direction})
		}
		findOpts.SetSort(sortDoc)
	}

	cursor, err := coll.Find(ctx, filter, findOpts)
	if err != nil {
		return nil, fmt.Errorf("go-dal: mongodb: find: %w", err)
	}
	defer cursor.Close(ctx)

	var results []map[string]interface{}
	if err := cursor.All(ctx, &results); err != nil {
		return nil, fmt.Errorf("go-dal: mongodb: find all: %w", err)
	}

	// Convert ObjectIDs to strings
	for i, doc := range results {
		if id, ok := doc["_id"].(primitive.ObjectID); ok {
			doc["_id"] = id.Hex()
		}
		results[i] = doc
	}

	return results, nil
}

// UpdateOne updates a single document matching the filter.
func (m *MongoDB) UpdateOne(ctx context.Context, collection string, filter, update interface{}) (int64, error) {
	coll := m.db.Collection(collection)

	result, err := coll.UpdateOne(ctx, filter, update)
	if err != nil {
		return 0, fmt.Errorf("go-dal: mongodb: update: %w", err)
	}

	return result.ModifiedCount, nil
}

// DeleteOne deletes a single document matching the filter.
func (m *MongoDB) DeleteOne(ctx context.Context, collection string, filter interface{}) (int64, error) {
	coll := m.db.Collection(collection)

	result, err := coll.DeleteOne(ctx, filter)
	if err != nil {
		return 0, fmt.Errorf("go-dal: mongodb: delete: %w", err)
	}

	return result.DeletedCount, nil
}

// Count returns the count of documents matching the filter.
func (m *MongoDB) Count(ctx context.Context, collection string, filter interface{}) (int64, error) {
	coll := m.db.Collection(collection)

	count, err := coll.CountDocuments(ctx, filter)
	if err != nil {
		return 0, fmt.Errorf("go-dal: mongodb: count: %w", err)
	}

	return count, nil
}

// CreateIndex creates an index on the collection.
func (m *MongoDB) CreateIndex(ctx context.Context, collection string, keys []dal.IndexKey, unique bool) error {
	coll := m.db.Collection(collection)

	sortDoc := bson.D{}
	for _, k := range keys {
		sortDoc = append(sortDoc, bson.E{Key: k.Field, Value: k.Direction})
	}

	indexModel := mongo.IndexModel{
		Keys: sortDoc,
	}

	indexOpts := options.Index().SetUnique(unique)
	indexModel.Options = indexOpts

	_, err := coll.Indexes().CreateOne(ctx, indexModel)
	if err != nil {
		return fmt.Errorf("go-dal: mongodb: create index: %w", err)
	}

	return nil
}

// Close closes the MongoDB connection.
func (m *MongoDB) Close(ctx context.Context) error {
	if err := m.conn.Disconnect(ctx); err != nil {
		return fmt.Errorf("go-dal: mongodb: disconnect: %w", err)
	}

	return nil
}
